admin/commands/storage/read_results.go

```diff
@@ -0,0 +1,119 @@
+package storage
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/onflow/flow-go/admin"
+	"github.com/onflow/flow-go/admin/commands"
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/state/protocol"
+	"github.com/onflow/flow-go/storage"
+)
+
+var _ commands.AdminCommand = (*ReadResultsCommand)(nil)
+
+type readResultsRequestType int
+
+const (
+	readResultsRequestByID readResultsRequestType = iota
+	readResultsRequestByBlock
+)
+
+type readResultsRequest struct {
+	requestType       readResultsRequestType
+	value             interface{}
+	numResultsToQuery uint64
+}
+
+type ReadResultsCommand struct {
+	state   protocol.State
+	results storage.ExecutionResults
+}
+
+func (r *ReadResultsCommand) Handler(ctx context.Context, req *admin.CommandRequest) (interface{}, error) {
+	data := req.ValidatorData.(*readResultsRequest)
+	var results []*flow.ExecutionResult
+	var resultID flow.Identifier
+	n := data.numResultsToQuery
+
+	switch data.requestType {
+	case readResultsRequestByID:
+		resultID = data.value.(flow.Identifier)
+	case readResultsRequestByBlock:
+		if header, err := getBlockHeader(r.state, data.value.(*blocksRequest)); err != nil {
+			return nil, fmt.Errorf("failed to get block header: %w", err)
+		} else if result, err := r.results.ByBlockID(header.ID()); err != nil {
+			return nil, fmt.Errorf("failed to get result by block ID: %w", err)
+		} else {
+			results = append(results, result)
+			resultID = result.PreviousResultID
+			n -= 1
+		}
+	}
+
+	for i := uint64(0); i < n && resultID != flow.ZeroID; i++ {
+		result, err := r.results.ByID(resultID)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get result by ID: %w", err)
+		}
+		results = append(results, result)
+		resultID = result.PreviousResultID
+	}
+
+	return convertToInterfaceList(results)
+}
+
+func (r *ReadResultsCommand) Validator(req *admin.CommandRequest) error {
+	input, ok := req.Data.(map[string]interface{})
+	if !ok {
+		return errors.New("wrong input format: expected JSON")
```

Comment:
we can move this and reuse this error message in all the Validator(req *admin.CommonRequest) funcs